rasterio/_vsiopener.pyx

```diff
@@ -0,0 +1,160 @@
+# Currently the only way to "create" a file in the filesystem is to add
+# an entry to this dictionary. GDAL will then Open the path later.
+cdef _OPENER_REGISTRY = {}
+cdef _OPEN_FILE_EXIT_STACKS = {}
+
+
+cdef int install_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    """Install handlers for python file openers if it isn't already installed."""
+    cdef char **registered_prefixes = VSIGetFileSystemsPrefixes()
+    cdef int prefix_index = CSLFindString(registered_prefixes, PREFIX_BYTES)
+    CSLDestroy(registered_prefixes)
+
+    if prefix_index < 0:
+        log.debug("Installing Python opener handler plugin...")
+        callbacks_struct = VSIAllocFilesystemPluginCallbacksStruct()
+        callbacks_struct.open = <VSIFilesystemPluginOpenCallback>pyopener_open
+        callbacks_struct.tell = <VSIFilesystemPluginTellCallback>pyopener_tell
+        callbacks_struct.seek = <VSIFilesystemPluginSeekCallback>pyopener_seek
+        callbacks_struct.read = <VSIFilesystemPluginReadCallback>pyopener_read
+        callbacks_struct.close = <VSIFilesystemPluginCloseCallback>pyopener_close
+        callbacks_struct.pUserData = <void*>_OPENER_REGISTRY
+        retval = VSIInstallPluginHandler(PREFIX_BYTES, callbacks_struct)
+        VSIFreeFilesystemPluginCallbacksStruct(callbacks_struct)
+        return retval
+    else:
+        return 0
+
+
+cdef void uninstall_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    if callbacks_struct is not NULL:
+        callbacks_struct.pUserData = NULL
+        VSIFreeFilesystemPluginCallbacksStruct(callbacks_struct)
+    callbacks_struct = NULL
+
+
+cdef void* pyopener_open(void *pUserData, const char *pszFilename, const char *pszAccess) with gil:
+    """Access files in the virtual filesystem.
+
+    This function is mandatory in the GDAL Filesystem Plugin API.
+    GDAL may call this function multiple times per filename and each
+    result must be seperately seekable.
+    """
+    if pszAccess != b"r" and pszAccess != b"rb":
+        log.error("Python opener is currently a read-only interface.")
+        return NULL
+
+    if pUserData is NULL:
+        log.error("Python opener registry is not initialized.")
+        return NULL
+
+    cdef dict registry = <object>pUserData
+    filename = pszFilename.decode("utf-8")
+
+    log.debug("Looking up opener: registry=%r, filename=%r", registry, filename)
+    try:
+        file_opener = registry[filename]
+    except KeyError:
+        log.info("Object not found: registry=%r, filename=%r", registry, filename)
+        return NULL
+
+    mode = pszAccess.decode("utf-8")
+    cdef object file_obj
+
+    try:
+        file_obj = file_opener(filename, mode)
+    # ZipFile.open doesn't accept binary modes like "rb" and will raise
+    # ValueError if given one. We strip the mode in this case.
+    except ValueError as err:
+        file_obj = file_opener(filename, mode.rstrip("b"))
```

Comment:
```suggestion
        file_obj = file_opener(filename, mode=mode.rstrip("b"))
```

Reply:
@snowman2 I'd like to adhere to the opener spec in https://docs.python.org/3/library/functions.html#open, which uses positional arguments.